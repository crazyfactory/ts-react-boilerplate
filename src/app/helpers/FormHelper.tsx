import * as React from "react";
import {WrappedFieldProps} from "redux-form";
import {style} from "typestyle";

interface IRenderFieldProps<T> extends WrappedFieldProps<T> {
  label?: string;
  type?: string;
}

export const required = (value) => (value ? undefined : "Required");

export const maxLength = (max) => (value) => value && value.length > max ? `Must be ${max} characters or less` : undefined;

export const minLength = (min) => (value) => value && value.length < min ? `Must be ${min} characters or more` : undefined;

export const numberType = (value) => value && isNaN(Number(value)) ? "Must be a number" : undefined;

export const minValue = (min) => (value) => value < min ? `Must be at least ${min}` : undefined;

export const matchedPwd = (value, otherValues) => value !== otherValues.password ? `Passwords not matched` : undefined;

export const email = (value) => value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value) ? "Invalid email address" : undefined;

export const tooOld = (value) => value && value > 65 ? "You might be too old for this" : undefined;

export const aol = (value) => value && /.+@aol\.com/.test(value) ? "Really? You still use AOL for your email?" : undefined;

/*tslint:disable:object-literal-sort-keys*/
const styles = {
  row: style({
    margin: "15px 0"
  }),
  inputItem: style({
    padding: 5,
    border: "1px solid #ccc"
  }),
  message: style({
    marginTop: 5
  }),
  error: style({
    color: "red"
  }),
  warning: style({
    color: "darkorange"
  })
};
/*tslint:enable:object-literal-sort-keys*/

/*tslint:disable:jsx-no-multiline-js*/
export const renderField = (props: IRenderFieldProps<any>) => {
  const {input, label, type, meta: {active, touched, error, warning}} = props;
  return (
    <div className={styles.row}>
      <label>{props.label}</label>
      <div>
        <input {...input} placeholder={label} type={type} className={styles.inputItem} />
        {
          (active || touched) &&
          (
            (error && <div className={styles.message}><span className={styles.error}>{error}</span></div>)
            ||
            (warning && <div className={styles.message}><span className={styles.warning}>{warning}</span></div>)
          )
        }
      </div>
    </div>
  );
};
