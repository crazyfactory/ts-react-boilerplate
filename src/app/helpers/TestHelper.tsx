/** React Specific */
import {mount, ReactWrapper, render, shallow, ShallowRendererProps, ShallowWrapper} from "enzyme";
import * as React from "react";
import {IntlProvider} from "react-intl";
import {Provider} from "react-redux";
import {createStore, Store} from "redux";
import rootReducer from "../redux/rootReducer";

/** Redux Mock Store Configuration */
import {IStore} from "../redux/IStore";
import {IState} from "../redux/modules/baseModule";
import {ILanguage} from "../redux/modules/languageModule";

declare type TComponent = React.ComponentClass<any> | React.SFC<any> | React.ClassType<any, any, any> | string;

export class TestHelper<TProps, TState> {

  protected state: Partial<IStore>;
  protected props: TProps;
  protected locale: string;
  protected languageData: object;
  private ComponentClass: TComponent;

  public withState(state: Partial<IStore>): TestHelper<TProps, TState> {
    this.state = state;
    return this;
  }

  public withProps(props: TProps): TestHelper<TProps, TState> {
    this.props = props;
    return this;
  }

  // sets translation data to the state AND does NOT set directly to intl provider
  public withTranslation(translation: ILanguage): TestHelper<TProps, TState> {
    const languageState: IState<ILanguage> = {payload: translation};
    this.state = {...this.state, language: languageState};
    return this;
  }

  public mount(component: TComponent): ReactWrapper<TProps, TState> {
    this.ComponentClass = component;
    return mount(
      this.getWrappedComponent()
    );
  }

  public shallow(component: TComponent, options?: ShallowRendererProps): ShallowWrapper<TProps, TState> {
    this.ComponentClass = component;
    return shallow(this.getWrappedComponent(), options);
  }

  public render(component: TComponent, renderOptions?: any): Cheerio {
    this.ComponentClass = component;
    return render(this.getWrappedComponent(), renderOptions);
  }

  private getWrappedComponent(): JSX.Element {
    const ComponentClass = this.ComponentClass;
    const providerComponent = (
      <Provider store={this.getStore()}>
        <ComponentClass {...this.props} />
      </Provider>
    );
    return this.state.language ? this.getWithTranslation(providerComponent) : providerComponent;
  }

  private getWithTranslation(component: JSX.Element): JSX.Element {
    return (
      <IntlProvider locale={this.state.language.payload.locale} messages={this.state.language.payload.languageData}>
        {component}
      </IntlProvider>
    );
  }

  private getStore(): Store<IStore> {
    return createStore<IStore>(rootReducer, (this.state as any));
  }

}
