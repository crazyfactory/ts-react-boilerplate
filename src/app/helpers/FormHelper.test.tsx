import {
  aol, email, matchedPwd, maxLength, minLength, minValue,
  numberType, renderField, required, tooOld
} from "../helpers/FormHelper";
import { TestHelper } from "../helpers/TestHelper";

describe("FormHelper", () => {

  describe("<renderField />", () => {
    it("without error and warning", () => {
      const props = {
        input: {},
        label: "username",
        meta: {
          active: false,
          error: false,
          touched: false,
          warning: false
        },
        type: "text"
      };
      const renderer = new TestHelper();
      const component = renderer.withProps(props).mount(renderField);
      expect(component).toMatchSnapshot();
    });

    it("with error when active or touched", () => {
      const props = {
        input: {},
        label: "username",
        meta: {
          active: true,
          error: true,
          touched: true,
          warning: false
        },
        type: "text"
      };
      const renderer = new TestHelper();
      const component = renderer.withProps(props).mount(renderField);
      expect(component).toMatchSnapshot();
    });

    it("with warning when active or touched", () => {
      const props = {
        input: {},
        label: "username",
        meta: {
          active: true,
          error: false,
          touched: true,
          warning: true
        },
        type: "text"
      };
      const renderer = new TestHelper();
      const component = renderer.withProps(props).mount(renderField);
      expect(component).toMatchSnapshot();
    });
  });

  describe("aol()", () => {
    it("invalidates aol email", () => {
      expect(aol("pai@aol.com")).toEqual("Really? You still use AOL for your email?");
    });

    it("validates email from other domains", () => {
      expect(aol("pai@pai.com")).toBeUndefined();
    });
  });

  describe("email()", () => {
    it("invalidates non email", () => {
      expect(email("abcd")).toEqual("Invalid email address");
    });
    it("validates email correctly", () => {
      expect(email("pai@pai.com")).toBeUndefined();
    });
  });

  describe("matchedPwd()", () => {
    it("invalidates unmatched passwords", () => {
      expect(matchedPwd("pass1", {password: "pass2"})).toEqual("Passwords not matched");
    });
    it("validates matched passwords", () => {
      expect(matchedPwd("pass", {password: "pass"})).toBeUndefined();
    });
  });

  describe("maxLength()", () => {
    it("invalidates too long string", () => {
      expect((maxLength(8)("123456789"))).toEqual("Must be 8 characters or less");
    });
    it("validates ok string", () => {
      expect((maxLength(8)("12345678"))).toBeUndefined();
    });
  });

  describe("minLength()", () => {
    it("invalidates too short string", () => {
      expect((minLength(8)("1234567"))).toEqual("Must be 8 characters or more");
    });
    it("validates ok string", () => {
      expect((minLength(8)("12345678"))).toBeUndefined();
    });
  });

  describe("minValue()", () => {
    it("invalidates too little value", () => {
      expect(minValue(5)(0)).toEqual("Must be at least 5");
    });

    it("validates ok value", () => {
      expect(minValue(5)(10)).toBeUndefined();
    });
  });

  describe("numberType()", () => {
    it("invalidates non number", () => {
      expect(numberType("abcd")).toEqual("Must be a number");
    });

    it("validates number", () => {
      expect(numberType(5)).toBeUndefined();
    });
  });

  describe("required()", () => {
    it("invalidates falsy value", () => {
      expect(required(0)).toEqual("Required");
    });

    it("validates truthy value", () => {
      expect(required("abcd")).toBeUndefined();
    });
  });

  describe("tooOld()", () => {
    it("invalidates number more than 65", () => {
      expect(tooOld(70)).toEqual("You might be too old for this");
    });

    it("validates number less than 65", () => {
      expect(tooOld(65)).toBeUndefined();
    });
  });

});
