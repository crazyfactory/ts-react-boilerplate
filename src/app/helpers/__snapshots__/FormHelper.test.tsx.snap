// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormHelper <renderField /> with error when active or touched 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Component
    input={Object {}}
    label="username"
    meta={
      Object {
        "active": true,
        "error": true,
        "touched": true,
        "warning": false,
      }
    }
    type="text"
  >
    <div
      className="fyjcn3x"
    >
      <label>
        username
      </label>
      <div>
        <input
          className="f17g2th6"
          placeholder="username"
          type="text"
        />
        <div
          className="f1w4o4fu"
        >
          <span
            className="f1jvcvsh"
          />
        </div>
      </div>
    </div>
  </Component>
</Provider>
`;

exports[`FormHelper <renderField /> with warning when active or touched 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Component
    input={Object {}}
    label="username"
    meta={
      Object {
        "active": true,
        "error": false,
        "touched": true,
        "warning": true,
      }
    }
    type="text"
  >
    <div
      className="fyjcn3x"
    >
      <label>
        username
      </label>
      <div>
        <input
          className="f17g2th6"
          placeholder="username"
          type="text"
        />
        <div
          className="f1w4o4fu"
        >
          <span
            className="fr6gy5q"
          />
        </div>
      </div>
    </div>
  </Component>
</Provider>
`;

exports[`FormHelper <renderField /> without error and warning 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Component
    input={Object {}}
    label="username"
    meta={
      Object {
        "active": false,
        "error": false,
        "touched": false,
        "warning": false,
      }
    }
    type="text"
  >
    <div
      className="fyjcn3x"
    >
      <label>
        username
      </label>
      <div>
        <input
          className="f17g2th6"
          placeholder="username"
          type="text"
        />
      </div>
    </div>
  </Component>
</Provider>
`;
